Add support for beta.mixcloud.com

joepvl opened the pull request:
As the title states. I could have sworn I saw an issue about this a few days ago, but I can't find it now. Anyway, Mixcloud is running a beta, and they've changed the selectors a bit. This PR adds support for the beta site/player to the Mixcloud controller. Of course, since it's a beta, the controller may need to be updated in the future as Mixcloud's engineers change things, but at least it works now if you've opted in.

Note that to get this to work on my machine I needed to apply the changes made in #105. I didn't want to wait until that one is merged before making this PR though, so here it is, based on the latest `master`. If #105 gets merged before this one, I will of course happily rebase this PR.

joepvl commented:
Also, I used the newly introduced `test` mechanism to provide multiple configs to `BaseController`. To be honest, I think it would be clearer to just manipulate the config you want to pass to the constructor based on the environment. That way the responsibility for the config stays within the file the controller is defined in.

msfeldstein commented on src/controllers/MixcloudController.js:
```diff
@@ -1,16 +1,37 @@
-    playpause: true,
-    previous: true,
-    next: true
+var isBeta = !!document.querySelector('.mz-player-control');
```
Will this ever be not true at the page load, but become true the second time test() is called?  Since it's cached here, it won't test it each time it tries, it will just store whatever the result is the first time this script is loaded

joepvl replied:
I may be misunderstanding something about how the extension works. Does it not only execute the controller code once, every time the page is loaded? If it does, then as far as I can tell this should always work. The page never switches between using the old and new players without reloading the page, and the element that is matched by the selector is served up in the initial payload in both cases (as in it is not rendered to the DOM through javascript from a template, it's already there in the html that is sent to the client).

Looking through the `BasicController` code, I can see that `test` is only ever called during class construction, and it's done synchronously. Unless that constructor is called multiple times during the lifecycle of the page, it doesn't even matter if the DOM changes so that another `test` would have returned `true` because the `test`s are never run more than once anyway! This is why I don't get the usefulness of the `test` mechanism 🙂 .

msfeldstein commented:
I see, i was misunderstanding, i thought we used `test()` in the init call, which is called multiple times until the page is ready to go.  I worry we'll need to do that with `test()` as well, imagine mixcloud has a preliminary 'loading' screen and we don't know which version you'll get until the rest of the page loads through ajax.  Can we make the isBeta test a function itself, rather than a cached value just for future proofing?

msfeldstein commented:
the workflow i was thinking was "initialize" will fail when the buttons it wants to observer aren't there yet, and succeed when it's found them and added its observers and stuff

joepvl commented:
Yeah I see where you're coming from, that makes more sense to me as well. I think what would be clearer here is to just put two `document.querySelector` calls into those test functions and call it a day for now, because I figure we'll need to update this controller once Mixcloud moves out of beta anyway. The code injection part has a specific selector too though, but we can change the injected code to make a similar decision based on what element is available in the page. I'll make these changes soon, hopefully today.

Side note: in this specific case, I would be surprised if they moved to for example having a loading screen and rendering from templates, because it looks like that would be a large architectural change for them and they don't seem to be investing in that area at the moment. At least as far as I can tell reasoning from the changes we're seeing now; they seem to still be using the same Angular 1 base in the beta as in the "old" site, the only changes being in the class names.

msfeldstein commented:
Yeah i don't think it really is an issue in this particular case, i think
it's just good to write things as callable functions instead of caching the
results ahead of time (for a hacky project like this which needs to be
flexible to change).  I've been burned by it before.

On Mon, Dec 12, 2016 at 4:08 PM Joep van Liempd <notifications@github.com>
wrote:

> Yeah I see where you're coming from, that makes more sense to me as well.
> I think what would be clearer here is to just put two
> document.querySelector calls into those test functions and call it a day
> for now, because I figure we'll need to update this controller once
> Mixcloud moves out of beta anyway. The code injection part has a specific
> selector too though, but we can change the injected code to make a similar
> decision based on what element is available in the page. I'll make these
> changes soon, hopefully today.
>
> Side note: in this specific case, I would be surprised if they moved to
> for example having a loading screen and rendering from templates, because
> it looks like that would be a large architectural change for them and they
> don't seem to be investing in that area at the moment. At least as far as I
> can tell reasoning from the changes we're seeing now; they seem to still be
> using the same Angular 1 base in the beta as in the "old" site, the only
> changes being in the class names.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/msfeldstein/chrome-media-keys/pull/106#issuecomment-266594406>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJ1b1UtyCw0So-Uc6f3FAz3FKgG7YUDks5rHeH1gaJpZM4LKeDe>
> .
>

joepvl commented:
Updated! From the main commit message:

> Made the Mixcloud controller more robust & clarified naming.
> The config tests now always check the page every time they are run.
> Also overhauled the in-page script so as to not leak the 'mxsway' object onto the page.
> Updated custom event code to not use a deprecated API.
> Lastly, simplified & updated parts of the code using ES2015 features.

msfeldstein commented on src/controllers/MixcloudController.js:
```diff
@@ -1,26 +1,48 @@
+
+const config = [
+  {
+    test: () => !isNewPlayer(),
```
In the case that we discussed, where there is a loading screen, this would pass even if it was a new player.  A better way to do it would be to have a positive case for both players, rather than depending on the other failing. 

Not a big deal for now, since as we said it doesn't matter too much for this particular case.
